README.md
# kalgan-config

Collection of functions to retrieve data and settings parameters defined in yaml files used by Kalgan Framework.

## Examples

This is the yaml file to be used in the following tests:
```yaml
## tests/settings.yaml

user:
  name: John
  is_real: false
  age: 39
  height: 1.78
  children:
    - Huey
    - Dewey
    - Louie
```
```rust
use kalgan_config::Config;

let config: Config = Config::new("tests/settings.yaml");
```
```rust
assert_eq!(config.get("user.name").unwrap(), "John");
```
```rust
assert_eq!(config.get_string("user.name").unwrap(), "John".to_string());
```
```rust
assert_eq!(config.get_bool("user.is_real").unwrap(), false);
```
```rust
assert_eq!(config.get_number("user.age").unwrap(), 39);
```
```rust
assert_eq!(config.get_float("user.height").unwrap(), 1.78);
```
```rust
assert_eq!(config.get_vec("user.children").unwrap(), vec!["Huey", "Dewey", "Louie"]);
```
    
## Documentation

For further information please visit:

* [Official Kalgan Site](https://kalgan.eduardocasas.com)
* [API Documentation on docs.rs](https://docs.rs/kalgan-config)


## License

This crate is licensed under either of the following licenses:

* [MIT License](https://choosealicense.com/licenses/mit/)
* [Apache License 2.0](https://choosealicense.com/licenses/apache-2.0/)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.9.1] - 2022-02-24
### Fixed
- Fix link to [API Documentation on docs.rs](https://docs.rs/kalgan-config).
- Only parse files with *.yaml* extension.
 
### Added
- Add this changelog.


src/parameters.rs
use log::{debug, error, info, trace, warn};
use serde::Deserialize;
pub use serde_yaml::Value;
use std::ffi::OsStr;
use std::{collections::HashMap, fs, path::Path};

pub(crate) fn generate(source: &str) -> crate::Config {
    info!("");
    info!("***********************************************************************************");
    info!("Crate: Config");
    info!("Start Config Parameters Generation");
    let mut config = crate::Config {
        collection: HashMap::new(),
    };
    if Path::new(source).exists() {
        if Path::new(&source).is_dir() {
            walk_folder_files(Path::new(&source), &mut config.collection);
        } else {
            get_file_content(Path::new(&source), &mut config.collection);
        }
    } else {
        error!("Source path {} not found.", source);
    }
    match config.collection.len() {
        0 => info!("Result: No parameters have been parsed"),
        1 => info!("Result: 1 parameter has been parsed"),
        _ => info!(
            "Result: {} parameters have been parsed",
            config.collection.len()
        ),
    }
    info!("End Config Parameters Generation");
    info!("***********************************************************************************");
    config
}
fn walk_folder_files(dir: &Path, parameters: &mut HashMap<String, Value>) {
    debug!("Reading folder {}...", dir.display());
    match fs::read_dir(dir) {
        Ok(read_dir) => {
            for entry in read_dir {
                let path = entry.unwrap().path();
                if path.is_dir() {
                    debug!("{} is dir", path.display());
                    walk_folder_files(&path, parameters);
                } else {
                    match path.extension().and_then(OsStr::to_str) {
                        Some(extension) => {
                            if extension.to_lowercase() == "yaml" {
                                debug!("{} is file", path.display());
                                get_file_content(&path, parameters);
                            } else {
                                debug!("{} is skipped.", path.display());
                            }
                        }
                        None => debug!("{} is skipped.", path.display()),
                    }
                }
            }
        }
        Err(e) => warn!("{}", e),
    }
}
fn get_file_content(path: &Path, parameters: &mut HashMap<String, Value>) {
    debug!("Reading file {}...", path.display());
    match fs::read_to_string(path) {
        Ok(config_file) => {
            let document = serde_yaml::Deserializer::from_str(&config_file);
            match Value::deserialize(document) {
                Ok(value) => iterate(value, parameters, &mut "".to_string(), ""),
                Err(e) => warn!("{}", e),
            };
        }
        Err(e) => warn!("{}", e),
    };
}
fn iterate(map: Value, parameters: &mut HashMap<String, Value>, subpath: &mut String, prev: &str) {
    for element in map.as_mapping() {
        for (index, value) in element.iter() {
            trace!("Parsing: {:?}", index);
            let key = match index.as_str() {
                Some(p) => p.to_string(),
                None => index.as_i64().unwrap().to_string(),
            };
            if key.contains(".") {
                warn!("Dots are not allowed in key names.");
                warn!("Parameter \"{}\" is skipped.", key);
                break;
            }
            let mut pos = match subpath.rfind(".") {
                Some(num) => num + 1,
                None => 0,
            };
            while !subpath.is_empty() && subpath[pos..].to_string() != prev {
                *subpath = if pos > 0 {
                    subpath[..pos - 1].to_string()
                } else {
                    "".to_string()
                };
                pos = match subpath.rfind(".") {
                    Some(num) => num + 1,
                    None => 0,
                };
            }
            if value.is_mapping() {
                if subpath.is_empty() {
                    subpath.push_str(format!("{}", key.clone()).as_str());
                } else {
                    subpath.push_str(format!(".{}", key.clone()).as_str());
                }
                iterate(value.clone(), parameters, subpath, &key);
            } else {
                let message_name = if subpath.is_empty() {
                    key.to_string()
                } else {
                    format!("{}.{}", subpath.clone(), key)
                };
                parameters.insert(message_name, value.clone());
            }
        }
    }
}


src/lib.rs
//! Collection of functions to retrieve data and settings parameters defined in yaml files.

pub use serde_yaml::Value;
use std::collections::HashMap;
mod parameters;

#[derive(Debug)]
/// The object that keeps the data collection.
///
/// This is the yaml file to be used in the following tests:
/// ```yaml
/// ## tests/settings.yaml
///
/// user:
///   name: John
///   is_real: false
///   age: 39
///   height: 1.78
///   children:
///     - Huey
///     - Dewey
///     - Louie
/// ```
pub struct Config {
    pub collection: HashMap<String, Value>,
}
impl Config {
    /// Creates and return the `Config` instance given the data source path (can be a file or a folder).
    /// # Examples:
    /// ```
    /// use kalgan_config::Config;
    ///
    /// let config: Config = Config::new("tests/settings.yaml");
    /// ```
    pub fn new(source: &str) -> Config {
        parameters::generate(source)
    }
    /// Returns the `serde_yaml::Value` for the given parameter.
    /// # Examples:
    /// ```
    /// # use kalgan_config::Config;
    /// # let config: Config = Config::new("tests/settings.yaml");
    /// assert_eq!(config.get("user.name").unwrap(), "John");
    /// ```
    pub fn get(&self, key: &str) -> Result<Value, String> {
        if self.exists(&key) {
            Ok(self.collection[key].clone())
        } else {
            Err(format!("Key \"{}\" not found.", &key))
        }
    }
    /// Returns the value as a `String` for the given parameter.
    /// # Examples:
    /// ```
    /// # use kalgan_config::Config;
    /// # let config: Config = Config::new("tests/settings.yaml");
    /// assert_eq!(config.get_string("user.name").unwrap(), "John".to_string());
    /// ```
    pub fn get_string(&self, key: &str) -> Result<String, String> {
        match self.get(&key)?.as_str() {
            Some(string) => Ok(string.to_string()),
            None => Err(format!("Value \"{}\" is not a string.", &key)),
        }
    }
    /// Returns the value as a `bool` for the given parameter.
    /// # Examples:
    /// ```
    /// # use kalgan_config::Config;
    /// # let config: Config = Config::new("tests/settings.yaml");
    /// assert_eq!(config.get_bool("user.is_real").unwrap(), false);
    /// ```
    pub fn get_bool(&self, key: &str) -> Result<bool, String> {
        match self.get(&key)?.as_bool() {
            Some(bool) => Ok(bool),
            None => Err(format!("Value \"{}\" is not boolean.", &key)),
        }
    }
    /// Returns the value as a `i64` for the given parameter.
    /// # Examples:
    /// ```
    /// # use kalgan_config::Config;
    /// # let config: Config = Config::new("tests/settings.yaml");
    /// assert_eq!(config.get_number("user.age").unwrap(), 39);
    /// ```
    pub fn get_number(&self, key: &str) -> Result<i64, String> {
        match self.get(&key)?.as_i64() {
            Some(num) => Ok(num),
            None => Err(format!("Value \"{}\" is not i64/u64.", &key)),
        }
    }
    /// Returns the value as a `f64` for the given parameter.
    /// # Examples:
    /// ```
    /// # use kalgan_config::Config;
    /// # let config: Config = Config::new("tests/settings.yaml");
    /// assert_eq!(config.get_float("user.height").unwrap(), 1.78);
    /// ```
    pub fn get_float(&self, key: &str) -> Result<f64, String> {
        match self.get(&key)?.as_f64() {
            Some(num) => Ok(num),
            None => Err(format!("Value \"{}\" is not f64.", &key)),
        }
    }
    /// Returns the value as a `Vec<serde_yaml::Value>` for the given parameter.
    /// # Examples:
    /// ```
    /// # use kalgan_config::Config;
    /// # let config: Config = Config::new("tests/settings.yaml");
    /// assert_eq!(config.get_vec("user.children").unwrap(), vec!["Huey", "Dewey", "Louie"]);
    /// ```
    pub fn get_vec(&self, key: &str) -> Result<Vec<Value>, String> {
        match self.get(&key)?.as_sequence() {
            Some(sequence) => Ok(sequence.clone()),
            None => Err(format!("Value \"{}\" is not a sequence.", &key)),
        }
    }
    pub fn exists(&self, key: &str) -> bool {
        self.collection.contains_key(&key.to_string())
    }
}
